Graph/Basic Graph Traversal/1462.py

```diff
@@ -0,0 +1,55 @@
+class Solution:
+    def checkIfPrerequisite(self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]) -> List[bool]:
+        ans = []
+        self.graph = defaultdict(list)
+
+        # Build an adjacency list to represent graph.
+        for a, b in prerequisites:
+            self.graph[a].append(b)
+
+        for u, v in queries:
+            ans.append(self.isPrerequisites(u, v))
+
+        return ans
+
+    # BFS Solution.
+    def isPrerequisites(self, start: int, end: int) -> bool:
+        queue = deque([start])
+        seen = set()
+
+        while queue:
+            curr = queue.popleft()
+
+            if curr == end:
+                return True
+
+            for v in self.graph[curr]:
+                if v not in seen:
+                    queue.append(v)
+                    seen.add(v)
+
+        return False
+
+    # DFS Solution
+    # Time Limited Exceeded.
```

Comment:
This solution gets TLE not because DFS (plus memoization) is inefficient. Rather, it is because there is a programming error in your code. I won't disclose too much for now; let's see if you can find this error yourself.

Reply:
I should recursively call function `dfs` instead of `isPrerequisites`.

However, both BFS/DFS solutions do not show a competitive time/space complexity in LC. Not sure if that is because the nature of the problem and the nature of Python, or there is more efficient way to do this problem.